src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/Razor/RazorDynamicFileInfoProvider.TextChangesTextLoader.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.LanguageServer.LanguageServer;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.HostWorkspace.Razor;
+
+internal partial class RazorDynamicFileInfoProvider
+{
+    private sealed class TextChangesTextLoader(
+        TextDocument? document,
+        IEnumerable<TextChange> changes,
+        byte[] checksum,
+        SourceHashAlgorithm checksumAlgorithm,
+        int? codePage,
+        Uri razorUri) : TextLoader
+    {
+        private readonly Lazy<SourceText> _emptySourceText = new Lazy<SourceText>(() =>
+        {
+            var encoding = codePage is null ? null : Encoding.GetEncoding(codePage.Value);
+            return SourceText.From("", checksumAlgorithm: checksumAlgorithm, encoding: encoding);
+        });
+
+        public override async Task<TextAndVersion> LoadTextAndVersionAsync(LoadTextOptions options, CancellationToken cancellationToken)
+        {
+            if (document is null)
+            {
+                var text = _emptySourceText.Value.WithChanges(changes);
+                return TextAndVersion.Create(text, VersionStamp.Default.GetNewerVersion());
+            }
+
+            var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+            // Validate the checksum information so the edits are known to be correct
+            if (IsSourceTextMatching(sourceText))
+            {
+                var version = await document.GetTextVersionAsync(cancellationToken).ConfigureAwait(false);
+                var newText = sourceText.WithChanges(changes);
+                return TextAndVersion.Create(newText, version.GetNewerVersion());
+            }
+
+            return await GetFullDocumentFromServerAsync(razorUri, cancellationToken).ConfigureAwait(false);
```

Comment:
I would say though we should evaluate this in a month or so - if we never get non-matching texts this should be changed into an assert, and if we do, we should track down where

Reply:
💯 should revisit to make sure this number is not high because that indicates perf is suboptimal